local-cluster/tests/local_cluster.rs

```diff
@@ -1680,7 +1601,282 @@ fn test_validator_saves_tower() {
     let tower4 = Tower::restore(&ledger_path, &validator_id).unwrap();
     trace!("tower4: {:?}", tower4);
     // should tower4 advance 1 slot compared to tower3????
-    assert_eq!(tower4.root(), tower3.root().map(|s| s + 1));
+    assert_eq!(tower4.root(), tower3.root() + 1);
+}
+
+fn last_vote_in_tower(ledger_path: &Path, node_pubkey: &Pubkey) -> Option<Slot> {
+    let tower = Tower::restore(&ledger_path, &node_pubkey);
+    if let Err(tower_err) = tower {
+        if tower_err.is_file_missing() {
+            return None;
+        } else {
+            panic!("tower restore failed...: {:?}", tower_err);
+        }
+    }
+    // actually saved tower must have at least one vote.
+    let last_vote = Tower::restore(&ledger_path, &node_pubkey)
+        .unwrap()
+        .last_voted_slot()
+        .unwrap();
+    Some(last_vote)
+}
+
+// A bit convoluted test case; but this roughly follows this test theoretical scenario:
+//
+// Step 1: You have validator A + B with 31% and 36% of the stake:
+//
+//  S0 -> S1 -> S2 -> S3 (A + B vote, optimistically confirmed)
+//
+// Step 2: Turn off A + B, and truncate the ledger after slot `S3` (simulate votes not
+// landing in next slot).
+// Start validator C with 33% of the stake with same ledger, but only up to slot S2.
+// Have `C` generate some blocks like:
+//
+// S0 -> S1 -> S2 -> S4
+//
+// Step 3: Then restart `A` which had 31% of the stake. With the tower, from `A`'s
+// perspective it sees:
+//
+// S0 -> S1 -> S2 -> S3 (voted)
+//             |
+//             -> S4 -> S5 (C's vote for S4)
+//
+// The fork choice rule weights look like:
+//
+// S0 -> S1 -> S2 (ABC) -> S3
+//             |
+//             -> S4 (C) -> S5
+//
+// Step 4:
+// Without the persisted tower:
+//    `A` would choose to vote on the fork with `S4 -> S5`. This is true even if `A`
+//    generates a new fork starting at slot `S3` because `C` has more stake than `A`
+//    so `A` will eventually pick the fork `C` is on.
+//
+//    Furthermore `B`'s vote on `S3` is not observable because there are no
+//    descendants of slot `S3`, so that fork will not be chosen over `C`'s fork
+//
+// With the persisted tower:
+//    `A` should not be able to generate a switching proof.
+//
+fn do_test_optimistic_confirmation_violation_with_or_without_tower(with_tower: bool) {
```

Comment:
wow, this looked really involved/painful 🤕. Thanks for implementing that hacky suggestion for a test, hopefully this makes us feel better about the optimistic confirmation + saved tower interaction!

Reply:
Yeah, to be frank, writing this was very hard and needed more logging improvement resulting in #12875. ;) Well, I'm actually surprised that I could write one. xD